evap/staff/tests/test_views.py

```diff
@@ -2617,3 +2620,29 @@ def test_make_other_semester_active(self):
 
         self.assertFalse(semester1.is_active)
         self.assertTrue(semester2.is_active)
+
+
+class TestStaffMode(WebTest):
```

Comment:
The `enter_staff_mode` view is has the correct access control decorator, and the middleware is written very defensively only ever settings permissions to False, but since this is security relevant, I think I could sleep better if we added a test here that, again, makes sure that:

If you are not a staff user and you access the "enter staff mode" url, you will get permission denied and you will not be in staff mode (aka you will not pass the is_staff check / will not be able to access a staff view / whatever).

Have you considered that?

Reply:
While I agree that this is important enough to warrant additional "defense layers", your suggestion would basically test the `staff_uncached_required` decorator. I mean you can do that, but there's really not much in there that can break.

What if we added an `assert user.has_staff_permission` in `update_staff_mode` in the middleware? though upper layers should of course handle that already, this would be an additional defense in the lowest layer and would protect not only against a potentially broken enter_staff_mode view.

or you do both :)

Reply:
> your suggestion would basically test the staff_uncached_required decorator

That is true, in my view it was mainly a "the decorator is there" test with the main intention of having it codified in two independent places (test and implementation)

> What if we added an assert user.has_staff_permission in update_staff_mode in the middleware? though upper layers should of course handle that already, this would be an additional defense in the lowest layer and would protect not only against a potentially broken enter_staff_mode view.

I would also be fine with that.

Reply:
This was a very good idea! As it turns out, until now the decorator allowed anyone in, because I forgot to mark `has_staff_permission` as a property, and `request.user.has_staff_permission` was truthy...
I fixed it, added the `assert` in the middleware, and added an extra test that a normal user gets 403 when trying to enter or exit staff mode.

Reply:
> but there's really not much in there that can break

I think I learned a valuable lesson today :D

Reply:
> As it turns out, until now the decorator allowed anyone in, because I forgot to mark has_staff_permission as a property, and request.user.has_staff_permission was truthy

That is such a bad bug and I think your only chance to spot that is a test. I guess in a statically typed language, linters should complain, but in python, pylint won't be able to figure out what we're talking about.

I had something similar recently, also concerning user access, where a code base used `def is_internal()` as a method on the `User` class, but stored `is_internal` as a boolean in a request handler `context`. All code used the context. When I wrote some new code that only had the user accessible, I made the same mistake. I wonder if there are any best practices to prevent this from happening. Do some people advocate using `== True` on truth-checks in security relevant code?